Cloud API: compose/id/metadata

achilleas-k opened the pull request:
This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
- [x] adequate documentation informing people about the change such as
  - [x] create a file in [news/unreleased](https://github.com/osbuild/osbuild-composer/tree/main/docs/news/unreleased) directory if this change should be mentioned in the release news
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

<!--
Thanks for proposing a change to osbuild-composer!

Please don't remove the above check list. These are things that each pull
request must have before it is merged. It helps maintainers to not forget
anything.

If the reason for ticking any of the boxes is ambiguous, please add a short
note explaining why.

For user-visible changes, "adequate documentation" is an entry describing the
change for users in docs/news. Please refer to docs/news/README.md for details.

In addition, if this pull request fixes a downstream issue, please refer to
test/README.md and add these additional items:

- [ ] 1st commit of any `rhbz#` related PR contains bug reproducer; CI reports FAIL or
- [ ] PR contains automated tests for new functionality and
- [ ] QE has approved reproducer/new tests and
- [ ] Subsequent commits provide bug fixes without modifying the reproducer; CI reports PASS and
- [ ] QE approves this PR; RHBZ status is set to `MODIFIED + Verified=Tested`
-->

_Followup to #1439_

Added a new Cloud API endpoint that returns metadata about a finished build.  Currently the metadata contains full package list (NEVRA) for the built image and, where available, the commit ID (hash) for the created Edge commit.

## Internal changes
- Utility functions for extracting and structuring the RPM metadata was moved to an internal package to be reused.
- Added a small utility function that reads the schema version from a raw Manifest.
- Metadata returned from stages was previously only stored for the RPM stage.  Now all metadata from all stages is stored.  Metadata from the `org.osbuild.ostree.commit` stage are also structured.

See individual commit messages for more details.

teg reviewed (dismissed):
Love it!

ondrejbudai commented on internal/distro/distro_test.go:
```diff
@@ -16,3 +23,41 @@ func TestDistro_Manifest(t *testing.T) {
 }
+
+func TestDistro_Version(t *testing.T) {
+	pipelinePath := "../../test/data/manifests/"
```
I wonder if we want to base this test on the actual testing manifests. They change quite frequently and with the migration of 8.5 and 9.0 manifests to v2, this test will probably need a rewrite. Maybe we want to consider using a fixed mock manifest inlined in Go? I think we already have several of those.

achilleas-k replied:
Yeah, that's definitely a better idea.

achilleas-k replied:
Done!

achilleas-k commented on internal/cloudapi/server.go:
```diff
@@ -521,3 +522,87 @@ func (server *Server) GetVersion(w http.ResponseWriter, r *http.Request) {
+		}
+		return
+	}
+	rpms = rpmmd.OSBuildStagesToRPMs(result.OSBuildOutput.Build.Stages)
```
This is wrong.  It's not the PRMs for the build root that we want but the RPMs for the payload.

achilleas-k replied:
I should also add a test for this.  A check to verify that a non-standard package specified in the request is included in the metadata.

achilleas-k replied:
Fixed.

achilleas-k commented:
Changed the result parsing so that it now supports collecting metadata from v2 manifest results as well.  It gets a little bit messy, but I'd like to clean it up in a separate PR where we switch to using v2 results internally (see #1493).

I tested the change against #1351 locally and I'll rebase that onto this PR once it's merged.  I also added a verification in the `api.sh` script to check if `postgresql` (specified in the request) is included in the package list in the metadata.

achilleas-k commented:
Don't believe GitLab CI report and click through instead.  Not sure why it's showing green, it hasn't properly started yet.

atodorov commented on internal/distro/distro_test.go:
```diff
@@ -16,3 +18,81 @@ func TestDistro_Manifest(t *testing.T) {
+		manifest := distro.Manifest(rawManifest)
+		detectedVersion, err := manifest.Version()
+		assert.NoError(err, "Could not detect Manifest version for %d: %v", idx, err)
+		assert.Equal(expectedVersion, detectedVersion, "in manifest %d", idx)
```
IMO in this case it would be better to use require.NoError (fails fast) instead of assert. Other than than the assertion functions are the same.

atodorov replied:
FTR in general with Go I prefer the `require` package over the `assert` one but I see we've used both of them.

achilleas-k replied:
I prefer letting it run through the rest of the tests even if one fails (where possible) since they're mostly independent and it's good to know if one or all of them are failing.  But it's not that important here.  Changed it to require.

achilleas-k commented:
I added a package verification step to all tests in `api.sh` but missed the fact that the Azure tests don't define any custom packages.

teg reviewed (approved):
Still great!
